### ISSUE TITLE
Make it easier to extend DocValuesTermsQuery (Jira: LUCENE-9026)

### ISSUE THREAD

**Issue by mocobeta** [mocobeta/sandbox-lucene-10557#193]
The visibility of some of the fields in DocValuesTermsQuery make it difficult to efficiently subclass.  Especially the "termData" instance variable, which is really core to the functioning of the class but is totally inaccessible from any sub-classes, forcing subclasses to store a duplicate PrefixCodedTerms object, and then juggle the state of both.

Are there any objections to making "termData" (and potentiall some other instance variables) protected instead of private for this class?

---
### Jira information

Original Jira: https://issues.apache.org/jira/browse/LUCENE-9026
Reporter: Jason Gerlowski
Assignee: Jason Gerlowski
Created: 2019-10-23T18:40:42.487+0000
Updated: 2019-12-29T20:49:45.814+0000
Resolved: 2019-11-11T20:56:38.798+0000

Issue Links:

Sub-Tasks:

Pull Requests:

**Comment by mocobeta** [comment]
What is your use-case for extending this query?

Author: Adrien Grand (`@jpountz`)
Created: 2019-10-23T21:09:49.687+0000
Updated: 2019-10-23T21:09:49.687+0000

**Comment by mocobeta** [comment]
I'm considering extending the query to implement Solr's [PostFilter](https://lucene.apache.org/solr/8_2_0/solr-core/org/apache/solr/search/PostFilter.html) interface.  The ultimate goal of this is to make it more efficient to run this query with very large numbers of terms.

Author: Jason Gerlowski
Created: 2019-10-24T12:29:04.974+0000
Updated: 2019-10-24T12:29:04.974+0000

**Comment by mocobeta** [comment]
I'm not familiar with two-phase intersection, though I'll look into it to see whether PostFilter is fully overlapped by that. 

I'm equally happy with a getter- if you have reason to prefer that I'll take my patch that direction.  Though I'm curious what difference you see between the two...Does the Lucene community steer users away from subclassing queries for a particular reason?  Or is there some more general aversion to `protected` visibility?

Author: Jason Gerlowski
Created: 2019-10-24T14:13:11.042+0000
Updated: 2019-10-24T14:13:11.042+0000

**Comment by mocobeta** [comment]
I agree with Adrien. In lots of places in Lucene implementation classes should be final, because if you want to change a query, add some filter or provide the functionality in an reuseable external class that can be used to compose a new query.

Author: Uwe Schindler (`@uschindler`)
Created: 2019-10-24T17:09:18.068+0000
Updated: 2019-10-24T17:09:18.068+0000

**Comment by mocobeta** [comment]
> I'm considering extending the query to implement Solr's [PostFilter](https://lucene.apache.org/solr/8_2_0/solr-core/org/apache/solr/search/PostFilter.html) interface. The ultimate goal of this is to make it more efficient to run this query with very large numbers of terms.
Hello, `@gerlowskija`. 

Please notice [method=docValuesTermsFilter](https://lucene.apache.org/solr/guide/7_3/other-parsers.html#terms-query-parser), I suppose it does it underneath.  
> **UPD**
> Ok. It can do that on lucene level, but including it as `fq` turns eager docset evaluation in SolrIndexSearcher.getProcessedFilter() thus it can only get gain as 
> ```java
> q={!bool filter='{!terms cache=false method=docValuesTermsFilter v=$termz}' must='' should=''}
> ```
 

Author: Mikhail Khludnev
Created: 2019-11-04T22:14:58.600+0000
Updated: 2019-11-05T07:33:34.326+0000

**Comment by mocobeta** [comment]
Thanks for the pointer Mikhail.  I looked at using `method=docValuesTermsFilter` for my use case, and it did do better than the default "method", but it's not the most efficient implementation possible.  (I have an alternate implementation that's done much better for us up for review on SOLR-13890, if you're curious.)

That said, making DVTQ more subclass/composition friendly is a generally good thing, regardless of whether we go forward with SOLR-13890.  So I plan on committing this small change later this evening. 

Author: Jason Gerlowski
Created: 2019-11-11T20:13:50.693+0000
Updated: 2019-11-11T20:13:50.693+0000
